cvat-ui/src/components/annotation-page/annotation-filter/annotation-filter-pane/annotation-filter-pane.tsx

```diff
@@ -0,0 +1,185 @@
+
+    const scrollFiltersToBottom = (): void => {
+        setTimeout(() => filtersEndRef?.current?.scrollIntoView({ block: 'center', behavior: 'smooth' }), 100);
+    };
+
+    const resetFilters = (e: React.MouseEvent<HTMLElement, MouseEvent>): void => {
+        e.preventDefault();
+        setFilters([]);
+    };
+
+    const confirmClearFilters = (e: React.MouseEvent<HTMLElement, MouseEvent>): void => {
+        e.preventDefault();
+        clearFiltersRef?.current?.click();
+    };
+
+    const isGroupingValid = (): boolean => {
+        let openBraces = 0;
+        let closeBraces = 0;
+        filters.forEach((filter: any) => {
+            openBraces += filter.left.length;
+            closeBraces += filter.right.length;
+        });
+        return openBraces === closeBraces;
+    };
+
+    useEffect(() => {
+        if (isGroupingValid()) {
+            const filtersStr = filtersPaneRef.current?.innerText.replace(/(?:\r\n|\r|\n)/g, '');
+            changeAnnotationsFilters(filters.length ? [filtersStr] : []);
+        }
+        scrollFiltersToBottom();
+    }, [filters]);
+
+    useEffect(() => {
+        if (!filterPanelVisible) setEditItem(undefined);
+    }, [filterPanelVisible]);
+
+    const addFilter = (filter: any): void => {
+        const newFilter = { ...filter };
+        newFilter.id = uuidv4();
+        newFilter.left = [];
+        newFilter.right = [];
+        setFilters([...filters, newFilter]);
+    };
+
+    const editFilter = (filter: any): void => {
+        let filterToEditIndex = -1;
+        filters.find((filterItem: any, filterItemIndex: number): boolean => {
+            filterToEditIndex = filterItemIndex;
+            return filterItem.id === filter.id;
+        });
+        if (filterToEditIndex >= 0) {
+            filters[filterToEditIndex] = { ...filter, id: uuidv4() };
+            setFilters([...filters]);
+        }
+        setFilterPanelVisible(false);
+    };
+
+    return (
+        <>
+            <Tooltip placement='topLeft' title={!isGroupingValid() ? 'Invalid grouping' : ''}>
+                <div
+                    ref={filtersPaneRef}
+                    className={`annotation-filters-pane ${isGroupingValid() ? '' : 'invalid'}`}
+                    onClick={() => {
+                        if (!filters.length) {
+                            setFilterPanelVisible(true);
+                        }
+                    }}
+                    style={{ cursor: filters.length ? 'default' : 'pointer' }}
+                    onContextMenu={(e: React.MouseEvent<HTMLElement, MouseEvent>) => confirmClearFilters(e)}
+                >
+                    {filters?.length ? (
+                        <>
+                            {filters.map((item: any) => (
+                                <AnnotationFilterItem
+                                    key={item.id}
+                                    item={item}
+                                    onGrouping={(filter: any) => {
```

Comment:
Please, avoid using ``any``, I believe it would be easy to specify precise types here